File: src/features/plugin/instalador.py
import os
import tempfile
from collections.abc import Callable, Mapping
from copy import deepcopy
from pathlib import Path
from typing import Protocol, cast

from src.features.plugin.adaptadores import AgentAdapter
from src.features.plugin.armazenamento import (
    atomic_write,
    dump_document,
    load_document,
    merge_mcp_entry,
    remove_mcp_entry,
)
from src.features.plugin.escopo import resolve_scope
from src.features.plugin.modelos import OperationResult, ScopeTarget
from src.features.plugin.skills import is_managed_skill, render_skill
from src.features.plugin.versoes import installed_version, latest_stable_version

_ENTRY_NAME = "compras-publicas-br"
_MISSING = object()
_CONFLICT = object()


def _mapping(value: object) -> Mapping[str, object] | None:
    if not isinstance(value, Mapping):
        return None
    return cast(Mapping[str, object], value)


class _VersionResolver(Protocol):
    def installed_version(self) -> str: ...

    def latest_stable_version(self) -> str: ...


VersionResolver = _VersionResolver | Callable[[], str]


class _TestingVersionResolver:
    def __init__(self, latest_version: str | None) -> None:
        self._latest_version = latest_version

    def installed_version(self) -> str:
        return installed_version()

    def latest_stable_version(self) -> str:
        return (
            self._latest_version
            if self._latest_version is not None
            else latest_stable_version()
        )


class InstallerService:
    def __init__(
        self,
        adapter: AgentAdapter,
        scope: ScopeTarget,
        version_resolver: VersionResolver,
        home: Path | None = None,
    ) -> None:
        self._adapter = adapter
        self._scope = scope
        self._version_resolver = version_resolver
        self._home = (home if home is not None else Path.home()).resolve()
        self._target = adapter.resolve_target(scope.scope, scope.root, self._home)
        adapter.validate_target(self._target)

    @classmethod
    def for_testing(
        cls, root: Path, agent: str, latest_version: str | None = None
    ) -> "InstallerService":
        from src.features.plugin.adaptadores import get_adapter

        resolved_scope = resolve_scope("project", root, home=root)
        return cls(
            get_adapter(agent),
            resolved_scope,
            _TestingVersionResolver(latest_version),
            root,
        )

    def install(self) -> OperationResult:
        version = self._installed_version()
        entry = self._adapter.build_entry(version)
        document, config_exists = self._read_config()
        existing = self._entry(document)
        if existing is _CONFLICT or (
            existing is not _MISSING and not self._adapter.owns_entry(existing)
        ):
            return self._result(
                changed=False,
                version=version,
                warnings=(self._conflict_warning(),),
            )

        merged, config_changed = merge_mcp_entry(
            document,
            self._target.container_path,
            _ENTRY_NAME,
            entry,
            self._adapter.owns_entry,
        )
        skill_content, skill_changed, skill_warning = self._planned_skill()
        if skill_warning is not None and not config_changed:
            return self._result(False, version, (skill_warning,))

        config_content = dump_document(merged, self._target.config_format)
        self._commit(
            config_content if config_changed or not config_exists else None,
            skill_content if skill_changed else None,
        )
        warnings = (skill_warning,) if skill_warning is not None else ()
        return self._result(config_changed or skill_changed, version, warnings)

    def update(self) -> OperationResult:
        document, config_exists = self._read_config()
        existing = self._entry(document)
        if existing is _CONFLICT:
            return self._result(False, "", (self._conflict_warning(),))
        if existing is _MISSING:
            return self._result(False, "", ("No managed MCP entry was found.",))
        if not self._adapter.owns_entry(existing):
            return self._result(False, "", (self._conflict_warning(),))

        version = self._latest_version()
        canonical_entry = self._adapter.build_entry(version)
        entry = self._update_pin(existing, canonical_entry)
        merged, config_changed = merge_mcp_entry(
            document,
            self._target.container_path,
            _ENTRY_NAME,
            entry,
            self._adapter.owns_entry,
        )
        skill_content, skill_changed, skill_warning = self._planned_skill()
        config_content = dump_document(merged, self._target.config_format)
        self._commit(
            config_content if config_changed or not config_exists else None,
            skill_content if skill_changed else None,
        )
        warnings = (skill_warning,) if skill_warning is not None else ()
        return self._result(config_changed or skill_changed, version, warnings)

    def uninstall(self) -> OperationResult:
        document, config_exists = self._read_config()
        existing = self._entry(document)
        managed_entry = existing is not _MISSING and existing is not _CONFLICT
        if managed_entry:
            managed_entry = self._adapter.owns_entry(existing)
        managed_skill = is_managed_skill(self._target.skill_path)
        warnings: list[str] = []
        if existing is _CONFLICT or (existing is not _MISSING and not managed_entry):
            warnings.append(self._conflict_warning())
        if self._has_unmanaged_skill():
            warnings.append(self._unmanaged_skill_warning())
        if not managed_entry and not managed_skill:
            warnings.append("No managed MCP entry or skill was found.")
        has_entry_conflict = existing is _CONFLICT or (
            existing is not _MISSING and not managed_entry
        )
        if has_entry_conflict and not managed_skill:
            return self._result(False, "", tuple(warnings))

        removed_document, config_changed = remove_mcp_entry(
            document,
            self._target.container_path,
            _ENTRY_NAME,
            self._adapter.owns_entry,
        )
        skill_changed = managed_skill
        version = self._entry_version(existing)
        config_content = dump_document(removed_document, self._target.config_format)
        self._commit(
            config_content if config_changed and config_exists else None,
            None,
            delete_skill=skill_changed,
        )
        return self._result(config_changed or skill_changed, version, tuple(warnings))

    def _read_config(self) -> tuple[dict[str, object], bool]:
        path = self._target.config_path
        if not path.exists():
            return {}, False
        return load_document(path, self._target.config_format), True

    def _entry(self, document: Mapping[str, object]) -> object:
        current: object = document
        for key in self._target.container_path:
            mapping = _mapping(current)
            if mapping is None:
                return _CONFLICT
            if key not in mapping:
                return _MISSING
            current = mapping[key]
        mapping = _mapping(current)
        if mapping is None:
            return _CONFLICT
        return mapping.get(_ENTRY_NAME, _MISSING)

    def _has_unmanaged_skill(self) -> bool:
        path = self._target.skill_path
        return (path.exists() or path.is_symlink()) and not is_managed_skill(path)

    def _planned_skill(self) -> tuple[str | None, bool, str | None]:
        path = self._target.skill_path
        if path.exists() or path.is_symlink():
            if not is_managed_skill(path):
                return None, False, f"Unmanaged skill preserved: {path}"
        content = render_skill(
            self._adapter.agent_id, self._adapter.display_name, self._scope.scope
        )
        try:
            unchanged = path.read_text(encoding="utf-8") == content
        except (FileNotFoundError, OSError, UnicodeError):
            unchanged = False
        return content, not unchanged, None

    def _commit(
        self,
        config_content: str | None,
        skill_content: str | None,
        delete_skill: bool = False,
    ) -> None:
        writes = tuple(
            (path, content)
            for path, content in (
                (self._target.config_path, config_content),
                (self._target.skill_path, skill_content),
            )
            if content is not None
        )
        snapshot_paths = tuple(path for path, _ in writes) + (
            (self._target.skill_path,) if delete_skill else ()
        )
        snapshots = {path: self._snapshot(path) for path in snapshot_paths}
        try:
            for path, _ in writes:
                path.parent.mkdir(parents=True, exist_ok=True)
            if config_content is not None:
                atomic_write(self._target.config_path, config_content)
            if skill_content is not None:
                atomic_write(self._target.skill_path, skill_content)
            elif delete_skill:
                self._target.skill_path.unlink()
        except BaseException:
            for path, snapshot in snapshots.items():
                self._restore(path, snapshot)
            raise

    def _snapshot(self, path: Path) -> tuple[bool, bytes | None]:
        exists = path.exists() or path.is_symlink()
        return exists, path.read_bytes() if exists else None

    def _restore(self, path: Path, snapshot: tuple[bool, bytes | None]) -> None:
        existed, content = snapshot
        if not existed:
            path.unlink(missing_ok=True)
        elif content is not None:
            atomic_write(path, content.decode("utf-8"))
            if path.read_bytes() != content:
                self._restore_bytes(path, content)

    def _restore_bytes(self, path: Path, content: bytes) -> None:
        temporary_path: Path | None = None
        try:
            with tempfile.NamedTemporaryFile(
                mode="wb",
                dir=path.parent,
                prefix=f".{path.name}.",
                suffix=".tmp",
                delete=False,
            ) as temporary:
                temporary_path = Path(temporary.name)
                temporary.write(content)
                temporary.flush()
                os.fsync(temporary.fileno())
            os.replace(temporary_path, path)
        finally:
            if temporary_path is not None:
                temporary_path.unlink(missing_ok=True)

    def _installed_version(self) -> str:
        if callable(self._version_resolver):
            return self._version_resolver()
        return self._version_resolver.installed_version()

    def _latest_version(self) -> str:
        if callable(self._version_resolver):
            return self._version_resolver()
        return self._version_resolver.latest_stable_version()

    def _update_pin(self, existing: object, canonical: Mapping[str, object]) -> dict[str, object]:
        entry = deepcopy(cast(dict[str, object], existing))
        existing_args = entry.get("args")
        canonical_args = canonical.get("args")
        if not isinstance(existing_args, list) or not isinstance(canonical_args, list):
            return dict(canonical)
        existing_args = cast(list[object], existing_args)
        canonical_args = cast(list[object], canonical_args)
        if len(existing_args) < 2 or len(canonical_args) < 2:
            return dict(canonical)
        existing_args[1] = canonical_args[1]
        return entry

    def _entry_version(self, entry: object) -> str:
        mapping = _mapping(entry)
        if mapping is not None:
            args = mapping.get("args")
            if not isinstance(args, list):
                return ""
            values = cast(list[object], cast(object, args))
            if len(values) > 1 and isinstance(values[1], str):
                return values[1]
        return ""

    def _conflict_warning(self) -> str:
        return f"Conflict preserved for managed MCP entry in {self._target.config_path}."

    def _unmanaged_skill_warning(self) -> str:
        return f"Unmanaged skill preserved: {self._target.skill_path}"

    def _result(self, changed: bool, version: str, warnings: tuple[str, ...]) -> OperationResult:
        return OperationResult(
            changed=changed,
            config_path=self._target.config_path,
            skill_path=self._target.skill_path,
            version=version,
            warnings=warnings,
        )

File: src/features/plugin/instalador.ts
import { randomBytes } from "node:crypto";
import {
  closeSync,
  existsSync,
  fsyncSync,
  lstatSync,
  mkdirSync,
  openSync,
  readFileSync,
  renameSync,
  rmSync,
  unlinkSync,
  writeSync,
} from "node:fs";
import { basename, dirname, join } from "node:path";

import type { AgentAdapter, AgentTarget } from "./adaptadores.js";
import {
  atomicWrite,
  dumpDocument,
  loadDocument,
  mergeMcpEntry,
  removeMcpEntry,
} from "./armazenamento.js";
import { isManagedSkill, renderSkill } from "./habilidades.js";
import type { OperationResult, ScopeTarget } from "./modelo.js";

const ENTRY_NAME = "compras-publicas-br";
const MISSING = Symbol("missing");
const CONFLICT = Symbol("conflict");

function asRecord(value: unknown): Record<string, unknown> | null {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

function isSymlink(path: string): boolean {
  try {
    return lstatSync(path).isSymbolicLink();
  } catch {
    return false;
  }
}

function pathExists(path: string): boolean {
  return existsSync(path) || isSymlink(path);
}

export type VersionResolver = () => string | Promise<string>;

export class InstallerService {
  private readonly _adapter: AgentAdapter;
  private readonly _scope: ScopeTarget;
  private readonly _versionResolver: VersionResolver;
  private readonly _target: AgentTarget;

  constructor(adapter: AgentAdapter, target: ScopeTarget, versionResolver: VersionResolver) {
    this._adapter = adapter;
    this._scope = target;
    this._versionResolver = versionResolver;
    this._target = adapter.resolveTarget(target);
    adapter.validateTarget(this._target);
  }

  async install(): Promise<OperationResult> {
    const version = await this._installedVersion();
    const entry = this._adapter.buildEntry(version);
    const [document, configExists] = this._readConfig();
    const existing = this._entry(document);
    if (
      existing === CONFLICT ||
      (existing !== MISSING && !this._adapter.ownsEntry(existing))
    ) {
      return this._result(false, version, [this._conflictWarning()]);
    }

    const [merged, configChanged] = mergeMcpEntry(
      document,
      this._target.containerPath,
      ENTRY_NAME,
      entry,
      (value: unknown) => this._adapter.ownsEntry(value)
    );
    const [skillContent, skillChanged, skillWarning] = this._plannedSkill();
    if (skillWarning !== null && !configChanged) {
      return this._result(false, version, [skillWarning]);
    }

    const configContent = dumpDocument(merged, this._target.configFormat);
    this._commit(
      configChanged || !configExists ? configContent : null,
      skillChanged ? skillContent : null
    );
    const warnings = skillWarning !== null ? [skillWarning] : [];
    return this._result(configChanged || skillChanged, version, warnings);
  }

  async update(): Promise<OperationResult> {
    const [document, configExists] = this._readConfig();
    const existing = this._entry(document);
    if (existing === CONFLICT) {
      return this._result(false, "", [this._conflictWarning()]);
    }
    if (existing === MISSING) {
      return this._result(false, "", ["No managed MCP entry was found."]);
    }
    if (!this._adapter.ownsEntry(existing)) {
      return this._result(false, "", [this._conflictWarning()]);
    }

    const version = await this._latestVersion();
    const canonicalEntry = this._adapter.buildEntry(version);
    const entry = this._updatePin(existing, canonicalEntry);
    const [merged, configChanged] = mergeMcpEntry(
      document,
      this._target.containerPath,
      ENTRY_NAME,
      entry,
      (value: unknown) => this._adapter.ownsEntry(value)
    );
    const [skillContent, skillChanged, skillWarning] = this._plannedSkill();
    const configContent = dumpDocument(merged, this._target.configFormat);
    this._commit(
      configChanged || !configExists ? configContent : null,
      skillChanged ? skillContent : null
    );
    const warnings = skillWarning !== null ? [skillWarning] : [];
    return this._result(configChanged || skillChanged, version, warnings);
  }

  async uninstall(): Promise<OperationResult> {
    const [document, configExists] = this._readConfig();
    const existing = this._entry(document);
    let managedEntry = existing !== MISSING && existing !== CONFLICT;
    if (managedEntry) {
      managedEntry = this._adapter.ownsEntry(existing);
    }
    const managedSkill = isManagedSkill(this._target.skillPath);
    const warnings: string[] = [];
    if (existing === CONFLICT || (existing !== MISSING && !managedEntry)) {
      warnings.push(this._conflictWarning());
    }
    if (this._hasUnmanagedSkill()) {
      warnings.push(this._unmanagedSkillWarning());
    }
    if (!managedEntry && !managedSkill) {
      warnings.push("No managed MCP entry or skill was found.");
    }
    const hasEntryConflict =
      existing === CONFLICT || (existing !== MISSING && !managedEntry);
    if (hasEntryConflict && !managedSkill) {
      return this._result(false, "", warnings);
    }

    const [removedDocument, configChanged] = removeMcpEntry(
      document,
      this._target.containerPath,
      ENTRY_NAME,
      (value: unknown) => this._adapter.ownsEntry(value)
    );
    const skillChanged = managedSkill;
    const version = this._entryVersion(existing);
    const configContent = dumpDocument(removedDocument, this._target.configFormat);
    this._commit(
      configChanged && configExists ? configContent : null,
      null,
      skillChanged
    );
    return this._result(configChanged || skillChanged, version, warnings);
  }

  private _readConfig(): [Record<string, unknown>, boolean] {
    const path = this._target.configPath;
    if (!existsSync(path)) {
      return [{}, false];
    }
    return [loadDocument(path, this._target.configFormat), true];
  }

  private _entry(document: Record<string, unknown>): unknown {
    let current: unknown = document;
    for (const key of this._target.containerPath) {
      const mapping = asRecord(current);
      if (mapping === null) return CONFLICT;
      if (!Object.hasOwn(mapping, key)) return MISSING;
      current = mapping[key];
    }
    const mapping = asRecord(current);
    if (mapping === null) return CONFLICT;
    return Object.hasOwn(mapping, ENTRY_NAME) ? mapping[ENTRY_NAME] : MISSING;
  }

  private _hasUnmanagedSkill(): boolean {
    const path = this._target.skillPath;
    return pathExists(path) && !isManagedSkill(path);
  }

  private _plannedSkill(): [string | null, boolean, string | null] {
    const path = this._target.skillPath;
    if (pathExists(path) && !isManagedSkill(path)) {
      return [null, false, `Unmanaged skill preserved: ${path}`];
    }
    const content = renderSkill(
      this._adapter.agentId,
      this._adapter.displayName,
      this._scope.scope
    );
    let unchanged: boolean;
    try {
      unchanged = readFileSync(path, "utf8") === content;
    } catch {
      unchanged = false;
    }
    return [content, !unchanged, null];
  }

  private _commit(
    configContent: string | null,
    skillContent: string | null,
    deleteSkill = false
  ): void {
    const target = this._target;
    const writes: [string, string][] = [];
    if (configContent !== null) writes.push([target.configPath, configContent]);
    if (skillContent !== null) writes.push([target.skillPath, skillContent]);
    const snapshotPaths = writes.map(([path]) => path);
    if (deleteSkill) snapshotPaths.push(target.skillPath);
    const snapshots = new Map(
      snapshotPaths.map((path) => [path, this._snapshot(path)] as const)
    );
    try {
      for (const [path] of writes) mkdirSync(dirname(path), { recursive: true });
      if (configContent !== null) atomicWrite(target.configPath, configContent);
      if (skillContent !== null) {
        atomicWrite(target.skillPath, skillContent);
      } else if (deleteSkill) {
        unlinkSync(target.skillPath);
      }
    } catch (error) {
      for (const [path, snapshot] of snapshots) this._restore(path, snapshot);
      throw error;
    }
  }

  private _snapshot(path: string): [boolean, Buffer | null] {
    const exists = pathExists(path);
    return [exists, exists ? readFileSync(path) : null];
  }

  private _restore(path: string, [existed, content]: [boolean, Buffer | null]): void {
    if (!existed) {
      rmSync(path, { force: true });
    } else if (content !== null) {
      atomicWrite(path, content.toString("utf8"));
      if (!readFileSync(path).equals(content)) this._restoreBytes(path, content);
    }
  }

  private _restoreBytes(path: string, content: Buffer): void {
    const tempPath = join(
      dirname(path),
      `.${basename(path)}.${randomBytes(8).toString("hex")}.tmp`
    );
    try {
      const fd = openSync(tempPath, "w");
      try {
        writeSync(fd, content);
        fsyncSync(fd);
      } finally {
        closeSync(fd);
      }
      renameSync(tempPath, path);
    } finally {
      rmSync(tempPath, { force: true });
    }
  }

  private async _installedVersion(): Promise<string> {
    return await Promise.resolve(this._versionResolver());
  }

  private async _latestVersion(): Promise<string> {
    return await Promise.resolve(this._versionResolver());
  }

  private _updatePin(
    existing: unknown,
    canonical: Record<string, unknown>
  ): Record<string, unknown> {
    const entry = structuredClone(existing) as Record<string, unknown>;
    const existingArgs = entry.args;
    const canonicalArgs = canonical.args;
    if (!Array.isArray(existingArgs) || !Array.isArray(canonicalArgs)) {
      return { ...canonical };
    }
    if (existingArgs.length < 2 || canonicalArgs.length < 2) {
      return { ...canonical };
    }
    existingArgs[1] = canonicalArgs[1];
    entry.args = existingArgs;
    return entry;
  }

  private _entryVersion(entry: unknown): string {
    const mapping = asRecord(entry);
    if (mapping !== null) {
      const args = mapping.args;
      if (!Array.isArray(args)) return "";
      if (args.length > 1 && typeof args[1] === "string") return args[1];
    }
    return "";
  }

  private _conflictWarning(): string {
    return `Conflict preserved for managed MCP entry in ${this._target.configPath}.`;
  }

  private _unmanagedSkillWarning(): string {
    return `Unmanaged skill preserved: ${this._target.skillPath}`;
  }

  private _result(
    changed: boolean,
    version: string,
    warnings: string[]
  ): OperationResult {
    return {
      changed: changed,
      configPath: this._target.configPath,
      skillPath: this._target.skillPath,
      version: version,
      warnings: warnings,
    };
  }
}
